### PR TITLE
Point cloud identify 3d

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description

Adds identification of 3D point cloud points:
The identification find all the points that intersect with the ray casted from the screen (points that are under the mouse cursor and are rendered).

https://user-images.githubusercontent.com/29183781/103883099-08c42980-50dd-11eb-80d3-ab30341667b6.mp4

### CODE REVIEW

**Comment by wonder-sk** [comment]
Great work @NEDJIMAbelgacem :tada: 

I have a couple of suggestions to make the code better, ready for future and easier to auto-test:

- ray representation and manipulation. Right now you are using two QVector3D values (origin+direction) to represent it, and operations with rays are scattered in multiple source files. I would like to suggest to create a new class (e.g. QgsRay3D) that would encapsulate those two vectors and they would also add a couple of methods:
  - "ray-box intersection" check (currently in __boxIntesects() in qgspointclouddataprovider.cpp)
  - "is point near ray" check (currently in getPointsOnRay() in qgspointclouddataprovider.cpp)

- transformation of screen position to a ray is inside Qgs3DMapCanvas::identifyPointCloudOnMouseEvent() - it would be good to extract it to a new static method, e.g. QgsRay3D Qgs3DUtils::rayFromScreenPoint( point, windowSize, camera )

- add auto-tests for the methods in the ray class, rayFromSceenPoint

- there is quite a lot of code duplication in Qgs3DMapToolIdentify::mouseReleaseEvent() that copies behavior of the 2D variant of identify tool. It will be difficult to maintain the two versions in sync. I would suggest that we update both 2D and 3D map tool in the following way: 1. the internal code for 2D and 3D variants will return identified points in some kind of intermediate representation (that could even be a vector of raw bytes) - you can create a new struct/class for this representation, 2. this intermediate representation would be passed to a static function in QgsMapToolIdentify that would convert it to the final identify results with texts and formatting

- getNodesIntersectingWithRay() function duplicates code of what the QgsChunkedEntity class does. I would like to avoid this duplication as well. I think Qgs3DMapToolIdentify::mouseReleaseEvent() should simply look at Qgs3DMapScene's layer entities, and for point cloud layers get indices of active nodes in their QgsChunkedEntity.

- my preference would be not to have code for identification in 3D inside QgsPointCloudDataProvider, Qgs3DMapCanvas, Qgs3DMapScene classes - better if we can keep most of the identification-specific code in Qgs3DMapToolIdentify

- there is some mixing of double/float arithmetics in QgsBox3d::distanceFromPoint() and Qgs3DMapScene::identifyPointCloudOnRay() - let's be very cautious about that - throughout QGIS we use 'double' values for coordinates because we have to: `float` only has precision of 6-9 significant decimal digits which is often not enough for map coordinates (which are often in orders of 10^8). We use floats only for 3D graphics because doubles are still poorly supported by GPUs - so we only lower the chance of numerical issues by using (0,0,0) at the middle of the scene in 3D world coordinates. Only saying this that a couple of times I have been badly bitten by numerical issues due to the usage of floats where I should have used doubles :-)

- a new struct CameraState is created in qgs3dmapscene.h but not used anywhere - remove?

File: src/core/geometry/qgsbox3d.cpp
Comment (nyalldawson) [inline src/core/geometry/qgsbox3d.cpp:128]: We need some unit tests for this too

File: src/core/geometry/qgsbox3d.cpp
Comment (NEDJIMAbelgacem) [reply src/core/geometry/qgsbox3d.cpp:133]: It does actually return 0 when the point is inside the box

File: src/core/geometry/qgsbox3d.h
Comment (nyalldawson) [inline src/core/geometry/qgsbox3d.h:211]: \since QGIS 3.18

File: src/core/geometry/qgsbox3d.cpp
Comment (nyalldawson) [inline src/core/geometry/qgsbox3d.cpp:121]: We still need some tests for this

File: src/core/geometry/qgsray3d.h
Comment (nyalldawson) [inline src/core/geometry/qgsray3d.h:1]: Needs copyright header

File: src/core/geometry/qgsray3d.h
Comment (nyalldawson) [inline src/core/geometry/qgsray3d.h:16]: Can you add some unit tests for this class?

File: src/core/geometry/qgsray3d.cpp
Comment (nyalldawson) [inline src/core/geometry/qgsray3d.cpp:1]: copyright header

File: src/3d/qgs3dmapscene.cpp
Comment (nyalldawson) [inline src/3d/qgs3dmapscene.cpp:1123]:
```suggestion
    if ( QgsPointCloudLayer *pc = qobject_cast<QgsPointCloudLayer *>( layer ) )
```

File: src/3d/qgs3dmapscene.cpp
Comment (nyalldawson) [inline src/3d/qgs3dmapscene.cpp:1128]:
```suggestion
      if ( !symbol )
```

File: src/app/3d/qgs3dmaptoolidentify.cpp
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolidentify.cpp:101]:
```suggestion
    QgsPointCloudLayer *pcLayer = qobject_cast< QgsPointCloudLayer * >( layerPoints[i].first );
```

File: src/core/pointcloud/qgspointclouddataprovider.cpp
Comment (nyalldawson) [inline src/core/pointcloud/qgspointclouddataprovider.cpp:460]:
```suggestion
      pointAttr[ tr( "Distance to camera" ) ] = ( point - ray.origin() ).length();
```

File: src/gui/qgsmaptoolidentify.h
Comment (nyalldawson) [inline src/gui/qgsmaptoolidentify.h:155]: I'd suggest marking this as SIP_SKIP -- it's not the kind of thing we want locked into the stable API pledge

File: tests/src/3d/testqgs3dutils.cpp
Comment (wonder-sk) [inline tests/src/3d/testqgs3dutils.cpp:182]: If possible at all, I would prefer that we do not create random numbers and do any extra math... Best tests are when we have some fixed numbers (they can be randomly chosen, but not during the runtime :-) ) and then fixed numbers which are the expected results... With no for-cycles, no extra math, basically to take any complexity away from tests, to lower the chance of bugs in the tests :-) ... and if we have let's say 2-3 different values that should be fine, maybe add some edge cases if you can think of some.

Also, with floating point numbers, I believe Qt uses fuzzy comparison in QCOMPARE, so it should be ok to just do QCOMPARE( a, b )

This would be an example of a better test:
```
QCOMPARE( ray.projectedPoint( QVector3D( 321, 654, 987 ) ), QVector3D( 123, 456, 789 ) );
```

Comment (NEDJIMAbelgacem) [reply tests/src/3d/testqgs3dutils.cpp]: I changed to the approach you suggested, I got a bit carried away with maths :-)
and somwhow QCOMPARE doesn't do fuzzy comarison properly for QVector3D so I ended up comparing components individually.

File: src/core/geometry/qgsray3d.h
Comment (NEDJIMAbelgacem) [reply src/core/geometry/qgsray3d.h:30]: Done

File: src/core/geometry/qgsbox3d.cpp
Comment (NEDJIMAbelgacem) [reply src/core/geometry/qgsbox3d.cpp:121]: Done

File: src/core/geometry/qgsbox3d.cpp
Comment (NEDJIMAbelgacem) [reply src/core/geometry/qgsbox3d.cpp]: Done

**Comment by NEDJIMAbelgacem** [comment]
Hi @wonder-sk and @nyalldawson 
I addressed the reviews on this PR and added some tests, if there is anything else I need to add please let me know

File: python/core/auto_generated/pointcloud/qgspointcloudattribute.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/pointcloud/qgspointcloudattribute.sip.in:82]: Can you hide this method from sip? We shouldn't be exposing the methods which deal with raw char pointers to python, as it has no way to obtain this data in the first place to have anything meaningful to call the method using...

**Comment by nyalldawson** [comment]
I can't see any issues here (apart from the sip note above) -- but I'll let @wonder-sk review before merging

File: python/core/auto_generated/pointcloud/qgspointcloudattribute.sip.in
Comment (NEDJIMAbelgacem) [reply python/core/auto_generated/pointcloud/qgspointcloudattribute.sip.in]: Sorry I just forgot to add SIP_SKIP

Comment (nyalldawson) [reply python/core/auto_generated/pointcloud/qgspointcloudattribute.sip.in]: Thanks!

**Review by wonder-sk — APPROVED** [review]
Looking great! :tada:

**Pull request merged**
